Window functions

CMJones001 opened the pull request:
Implementation of #179 allowing for more complex actions a window, by giving a command key before selecting the window.

Currently, this adds `x` to kill the window and `q` to toggle floating, although this can be extended to much more complex functions. A _full screen_ command and a _split and launch [x]_ seem like good next candidates. 

The main points:

- We add a `WindowCommand` Enum, defining the commands through `WindowCommand.send_to_window()`. If the user gives a key in the `key : WindowCommand` HashMap `window_cmd_map` then we select that command, otherwise we default to `WindowCommand::Focus`. 

- The command keys are currently hard-coded, but if you're happy with the implementation, I'll add something to the command line args. `--kill-window-char [x]` and `--focus-window-char [x]` are pretty verbose, but seem to be the simplest way of allowing the user to configure the keys?

- The Emacs packages used as reference silently ignores any keys used for commands, if they are also present in the `hint_chars`, however, I've decided to make this explicitly return an error, which seems more in line with Rust.


Happy for any other feedback!

svenstaro reviewed (changes requested):
Clean and simple, I like it! Great feature, too! I'd like for the keys to be customizable the way you already mentioned. Please also document this stuff in the README and write a small blip for the CHANGELOG. :)

svenstaro commented on src/utils.rs:
```diff
@@ -2,6 +2,7 @@ use anyhow::{bail, Context, Result};
 use itertools::Itertools;
 use log::debug;
 use regex::Regex;
+use std::collections::HashMap;
 use std::ffi::CStr;
 use std::iter;
 use std::thread::sleep;
```
While you're here: Could you order `std` stuff before other stuff?

svenstaro commented on src/utils.rs:
```diff
@@ -67,6 +68,25 @@ pub fn extents_for_text(text: &str, family: &str, size: f64) -> Result<cairo::Te
 }
 
+/// Return an error if there is a collision between the keys in the window function map and the hint
+/// chars
```
```suggestion
/// chars.
```

svenstaro commented on src/utils.rs:
```diff
@@ -67,6 +68,25 @@ pub fn extents_for_text(text: &str, family: &str, size: f64) -> Result<cairo::Te
+) -> Result<()> {
+    let colliding_keys: String = window_cmd_map
+        .keys()
+        .copied()
```
Do you really need this `.copied()`? I think a reference will probably be fine here for the filter.

CMJones001 replied:
I had some trouble with this, if we don't we end up with an iterator of double referenced `&&str` otherwise. `.map(|k| *k)` before the `collect()` also works?

svenstaro replied:
Eh in that case the copy is fine.

svenstaro commented on src/main.rs:
```diff
@@ -64,6 +66,15 @@ fn main() -> Result<()> {
 
+    // TODO: Move the ``window_cmd_map`` into the configuration
+    let mut window_command = &wm::WindowCommand::Focus;
+    let mut window_cmd_map = HashMap::new();
+    window_cmd_map.insert("x", wm::WindowCommand::Kill);
+    window_cmd_map.insert("q", wm::WindowCommand::Float);
```
Let's get rid of this `mut` and us the `HashMap::form([...])` syntax.

CMJones001 replied:
Done, but I think it'll get made redundant when I add the cofiguration

svenstaro commented on src/main.rs:
```diff
@@ -64,6 +66,15 @@ fn main() -> Result<()> {
     ];
 
+    // TODO: Move the ``window_cmd_map`` into the configuration
+    let mut window_command = &wm::WindowCommand::Focus;
```
Hm I don't really like much that we have this `mut` command here as there's quite a bit of mental burden associated with this giant function and this variable could change at any time in this long block of code. Could you try to re-engineer that? Perhaps by making use of expressions in the block below where you default to `Focus` if no other key is found in the sequence.

svenstaro commented on src/main.rs:
```diff
@@ -265,15 +276,19 @@ fn main() -> Result<()> {
                         if app_config.hint_chars.contains(kstr) {
                             info!("Adding '{}' to key sequence", kstr);
                             pressed_keys.push_str(kstr);
+                        } else if window_cmd_map.contains_key(kstr) {
+                            window_command = window_cmd_map.get(kstr).unwrap();
+                            info!("Changing window_command to '{window_command:?}'");
+                            continue;
```
Wouldn't this cause the sequence `fxf` to change the command to `Kill` in the middle of a focus sequence if there's a window with the label `fa`? Maybe I'm misunderstanding the code but I think we should require to the special code to be at the beginning of the sequence in order to be considered.

CMJones001 replied:
Yeah, I did just side step most of the other logic, I'll see how well I can get them to play together...

CMJones001 commented:
This should cover most of your comments, in particular, merging the logic of the window selection and command selection better. 

Now we add the command character to the `pressed_keys` if it is pressed at the start of the sequence (ignored otherwise) and the command selection logic is moved until later, so we can remove the extra `mut`.  

Maybe `pressed_keys` could do with being refactored to a struct and moving the command keys to a field rather than doing string parsing? But, this is the most I can do for now, before the weekend.

CMJones001 commented:
I've added configuration options for `--kill-window-char` and `--float-window-char`. I'm not too familiar with Clap, so I'm not sure if there's a more elegant way of doing things. 

The README and CHANGELOG are also updated :)
